### PR TITLE
Transparent distributed model training through TensorLayer GPU Trainer

### PR DESCRIPTION
**Pull request by luomai**
### Checklist
- [x] I've tested that my changes are compatible with the latest version of Tensorflow.
- [x] I've read the [Contribution Guidelines](https://github.com/tensorlayer/tensorlayer/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
By far the distributed GPU training of TensorLayer is still a difficult to configure given its dependency to the Worker and Parameter Server configuration in TensorFlow.

In the proposed Horovod trainer, we provide a simple training interface that a user only needs to provide:

- A model function that takes a X (i.e., example, sample) and Y_ (i.e., label) as inputs, and returns a loss (i.e., cost).
- A TensorFlow dataset that takes the training data

The user does not change any hyper-parameter of their models and does not to care about cluster configuration. TensorLayer and Horovod handle all the details underneath for the users.

### Description

This PR would require adding a horovod-based TL trainer in the distributed module, and remove deprecated worker-ps module, and add a CLI command to help users easily configure a Linux server to support horovod.

Logging TODO:

We are having two warning messages:
- [x] https://github.com/uber/horovod/issues/300
- [x] WARNING:tensorflow:It seems that global step (tf.train.get_global_step) has not been increased. 

The second message is produced as the validation step does not touch the optimizer which increases the global step.

API TODO:

- [x] Add an option to direct logs to a single place that helps debugging. This single place is likely to be TensorDB in a multi-node environment.
- [x] Add an option to disable "tfevents" if TensorBoard is not used.

Test TODO: 
- [x] Fix doc test
- [x] Fix RTD build

@lgarithm @zsdonghao

### CODE REVIEW

**Comment by zsdonghao** [comment]
The API is not flexible enough, there are many cases need to be considered.

DCGAN : X only, no Y
            2 losses
GAN-CLS: X1, X2 --> Y
            3 losses

**Comment by zsdonghao** [comment]
This is the API I would imagine.

```python
x, y = dataset

net = model(x)

loss = tl.cost….(net.outputs…)

lr = 0.0001
global_step = tf.contrib.framework.get_or_create_global_step()
#with tf.variable_scope('learning_rate'):
#      lr_v = tf.Variable(lr, trainable=False)
lr_v = tf.train.exponential_decay(
    learning_rate=lr,
    global_step=global_step,
    decay_steps=100,
    decay_rate=0.5,
    staircase=False,
    name=None
)
train_opt = tf.train.Optimizer(learning_rate=lr_v,  loss)

trainer = Horovod(train_opt, [x, y])

#for i in range(n_epoch):
     # i may update learning rate here
      # sess.run(tf.assign(lr_v, lr * 0.5))
     # update 1 iternation	
while not trainer.mon_sess.should_stop():
     trainer.update()
     print(“global steps:%d learning rate:%d” % (sess.run(global_step), sess.run(lr_v)))
     if trainer.is_master:
          tl.files.save_npz(net, ‘mode.npz’)
```

File: tensorlayer/distributed.py [hunk 1]
```diff
@@ -8,7 +8,79 @@
...
 
-__all__ = ['TaskSpecDef', 'TaskSpec', 'DistributedSession', 'StopAtTimeHook', 'LoadCheckpoint']
+import horovod.tensorflow as hvd
+
+from . import utils
+
+__all__ = ['TaskSpecDef', 'TaskSpec', 'DistributedSession', 'StopAtTimeHook', 'LoadCheckpoint', 'SimpleTrainer']
+
+
+class SimpleTrainer(object):
```
Comment (zsdonghao) [inline tensorlayer/distributed.py]: HorovodSimpleTrainer ?

File: example/tutorial_mnist_horovod.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        return network, tl.cost.cross_entropy(network.outputs, y_, name='cost')
+
+
+# load mnist data
+X_train, y_train, X_val, y_val, X_test, y_test = tl.files.load_mnist_dataset(shape=(-1, 784))
+dataset = make_dataset(X_train, y_train)
+
+optimizer = tf.train.RMSPropOptimizer(0.001)
+trainer = tl.distributed.HorovodTrainer(model_function, dataset, optimizer)
+trainer.run()
```
Comment (zsdonghao) [inline example/tutorial_mnist_horovod.py]: while not trainer.sess.should_stop():
      trainer.train_batch(self)
      if trainer.is_master and (sess.run(global_step) % 10000 == 0):
            tl.files.save_npz(trainer.net, ‘mode.npz’)

File: example/tutorial_mnist_horovod.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+trainer = tl.distributed.SimpleTrainer(model_function, dataset, optimizer)
+
+# start training
+trainer.train_to_end()
+# alternatively, you may want to control the training process
+import time
+while not trainer.sess.should_stop():
+    start_time = time.time()
+    # run 1 iteration 
+    err = trainer.train_batch()
```
Comment (zsdonghao) [inline example/tutorial_mnist_horovod.py]: train_batch need to return the loss of every iternation

**Comment by zsdonghao** [comment]
@luomai @lgarithm it seems the dataset part missing shuffle .

File: tensorlayer/distributed.py [hunk 4]
```diff
@@ -7,10 +7,93 @@
...
+
+        # The MonitoredTrainingSession takes care of session initialization,
+        # restoring from a checkpoint, saving to a checkpoint, and closing when done
+        # or an error occurs.
+        self.sess = tf.train.MonitoredTrainingSession(checkpoint_dir=checkpoint_dir, hooks=hooks, config=config)
+
+    def train_batch(self):
+        self.sess.run(self._train_op)
+
+    def validate_batch(self):
```
Comment (zsdonghao) [inline tensorlayer/distributed.py]: if validate_dataset is None, print warning?

File: example/tutorial_mnist_distributed_trainer.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+# load mnist data
+X_train, y_train, X_val, y_val, X_test, y_test = tl.files.load_mnist_dataset(shape=(-1, 784))
+training_dataset = make_dataset(X_train, y_train)
+validation_dataset = make_dataset(X_val, y_val)
+test_dataset = make_dataset(X_test, y_test)
+
+trainer = tl.distributed.DistributedTrainer(
+    network_and_cost_func=make_network, training_dataset=training_dataset, validation_dataset=validation_dataset
+    optimizer=tf.train.RMSPropOptimizer, optimizer_args={'learning_rate': 0.001}
```
[the remark this answers is not in the document]
Comment (zsdonghao) [reply example/tutorial_mnist_distributed_trainer.py]: `tl.distributed.DistributedTrainer` is wired..  as two `distributed` repeat.. I suggest to change it back to tl.distributed.SimpleTrainer`

**Comment by lgarithm** [comment]
The program creates a large (1.3G) file `events.out.tfevents.*` on start, we need to understand what it is for.

**Comment by lgarithm** [comment]
It looks like the size of `events.out.tfevents.*` being too large is caused by dataset.
When training without dataset, it is just around 5MB.

**Comment by luomai** [comment]
Are you sure you replicate the same behaviours as DataSet? DataSet is shuffling, batching and repeating the data. These operations can take extra space.

**Comment by lgarithm** [comment]
I can understand that dataset may take extra space, but I don't expect it to be that large.
The original size of mnist data is only 53MB (unzipped).

Another question is, why data should be in checkpoint?
Shouldn't checkpoint only contain parameters?

**Comment by luomai** [comment]
The dataset is not checkpointed. Only parameters are checkpointed.

**Comment by lgarithm** [comment]
Then the checkpoint folder should be small, the size of our model is about 

```
 (28*28 * 800 + 800*800 + 800 * 10) * 4 / 1024 / 1024 = 4  # MB
```

So what's inside `events.out.tfevents.*`?

**Comment by luomai** [comment]
We are having two warning messages:
- https://github.com/uber/horovod/issues/300
- WARNING:tensorflow:It seems that global step (tf.train.get_global_step) has not been increased. 

The second message is produced as the validation step does not touch the optimizer which increases the global step.

**Comment by zsdonghao** [comment]
I think this API can be merged, and mark as alpha version

**Comment by luomai** [comment]
We still need to write down the Document and fix some todo before merging.

**Comment by DEKHTIARJonathan** [comment]
@zsdonghao I don't think we should approve a PR when it's not ready to merge, we could indivertibly click on merge ... That would be bad ... Travis build is not passing yet

**Comment by lgarithm** [comment]
Yes, this is an attempt to resolve that, hope it would work.

**Comment by luomai** [comment]
Awesome, crossed finger. :)

**Comment by DEKHTIARJonathan** [comment]
I'm not sure that a script to install CUDA and CuDNN should be added in the repo (2bc6094), the versions are updated very oftenly, the installation process changes if depending on the OS you use.

I would think a CUDA_CuDNN_Install.md is much more appropriate ;)

If we add a script to do this, we will constently have issues telling us, it's not working anymore or outdated.

**Comment by luomai** [comment]
@DEKHTIARJonathan This is a script I add *temporarily* for helping some students to setup the environment to test this PR. It would not be merged into the master.

**Comment by zsdonghao** [comment]
for integrating with tensordb:

https://github.com/tensorlayer/tensorlayer/pull/700/files#diff-85b60e70886c2163c17af1edf5566e76

@luomai @fangde @OwenLiuzZ

**Comment by luomai** [comment]
@lgarithm can you have a final look and do the merging. It looks good to me.

@zsdonghao @OwenLiuzZ @fangde we will do the TensorDB integration later.
